stores/yaml/store.go

```diff
@@ -139,62 +291,96 @@ func (store *Store) LoadEncryptedFile(in []byte) (sops.Tree, error) {
-// sops.Tree runtime obejct
+// sops.Tree runtime object
 func (store *Store) LoadPlainFile(in []byte) (sops.TreeBranches, error) {
-	var data []yaml.MapSlice
-	if err := (yaml.CommentUnmarshaler{}).UnmarshalDocuments(in, &data); err != nil {
-		return nil, fmt.Errorf("Error unmarshaling input YAML: %s", err)
-	}
-
 	var branches sops.TreeBranches
-	for _, doc := range data {
-		branches = append(branches, store.mapSliceToTreeBranch(doc))
+	d := yaml.NewDecoder(bytes.NewReader(in))
+	for true {
+		var data yaml.Node
+		err := d.Decode(&data)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("Error unmarshaling input YAML: %s", err)
+		}
+
+		branch, err := store.yamlDocumentNodeToTreeBranch(data)
+		if err != nil {
+			return nil, fmt.Errorf("Error unmarshaling input YAML: %s", err)
+		}
+		branches = append(branches, branch)
 	}
 	return branches, nil
 }
 
 // EmitEncryptedFile returns the encrypted bytes of the yaml file corresponding to a
 // sops.Tree runtime object
 func (store *Store) EmitEncryptedFile(in sops.Tree) ([]byte, error) {
-	out := []byte{}
-	for i, branch := range in.Branches {
-		if i > 0 {
-			out = append(out, "---\n"...)
-		}
-		yamlMap := store.treeBranchToYamlMap(branch)
-		yamlMap = append(yamlMap, yaml.MapItem{Key: "sops", Value: stores.MetadataFromInternal(in.Metadata)})
-		tout, err := (&yaml.YAMLMarshaler{Indent: 4}).Marshal(yamlMap)
+    var b bytes.Buffer
+	e := yaml.NewEncoder(io.Writer(&b))
+	e.SetIndent(4)
+	for _, branch := range in.Branches {
+		// Document root
+		var doc = yaml.Node{}
+		doc.Kind = yaml.DocumentNode
+		// Add global mapping
+		var mapping = yaml.Node{}
+		mapping.Kind = yaml.MappingNode
+		doc.Content = append(doc.Content, &mapping)
+		// Create copy of branch with metadata appended
+		branch = append(sops.TreeBranch(nil), branch...)
+		branch = append(branch, sops.TreeItem{
+			Key: "sops",
+			Value: stores.MetadataFromInternal(in.Metadata),
+		})
+		// Marshal branch to global mapping node
+		store.appendTreeBranch(branch, &mapping)
+		// Encode YAML
+		err := e.Encode(&doc)
 		if err != nil {
 			return nil, fmt.Errorf("Error marshaling to yaml: %s", err)
 		}
-		out = append(out, tout...)
 	}
-	return out, nil
+	e.Close()
+	return b.Bytes(), nil
 }
 
 // EmitPlainFile returns the plaintext bytes of the yaml file corresponding to a
 // sops.TreeBranches runtime object
 func (store *Store) EmitPlainFile(branches sops.TreeBranches) ([]byte, error) {
-	var out []byte
-	for i, branch := range branches {
-		if i > 0 {
-			out = append(out, "---\n"...)
+    var b bytes.Buffer
+	e := yaml.NewEncoder(io.Writer(&b))
+	e.SetIndent(4)
+	for _, branch := range branches {
+		// Document root
+		var doc = yaml.Node{}
+		doc.Kind = yaml.DocumentNode
+		// Add global mapping
+		var mapping = yaml.Node{}
+		mapping.Kind = yaml.MappingNode
+		// Marshal branch to global mapping node
+		store.appendTreeBranch(branch, &mapping)
+		if len(mapping.Content) == 0 {
```

Comment:
Could you tell me what this does?
This looks cause #907 , and removing this looks cause nothing bad (as I don't know what good this does).

Reply:
Created #908

Reply:
It is needed so that empty document works well with https://github.com/go-yaml/yaml/pull/690. Unfortunately that fix for yaml.v3 isn't progressing, so I guess for now we can remove that branch. Once that fix landed in yaml.v3, this branch is definitely needed though.

Reply:
I tested the behavior of `sops` with your implementation of yaml.v3 by replacing in go.mod:
```
replace gopkg.in/yaml.v3 v3.0.0-20210107172259-749611fa9fcc => github.com/felixfontein/yaml v0.0.0-20210209202929-35d69a41298b
```
(35d69a41298b is the last commit of https://github.com/felixfontein/yaml/tree/improve-empty-document-handling)

It results `sops` confuse "no document" and "empty mapping":
```
# echo -e "# no document\n---\n# empty mapping\n{}\n" | sops --input-type yaml --output-type yaml -e /dev/stdin | sops --input-type yaml --output-type yaml -d /dev/stdin
# no document
---
# empty mapping
```

They are distinguished in yaml: http://ben-kiki.org/ypaste/data/21512/index.html

Unfortunately, the internal structure of `sops` (`TreeBranch`) cannot distinguish "no document" and "empty mapping".
I think we need to extend `TreeBranch` to have a flag that indicates "no document".

Reply:
> It results `sops` confuse "no document" and "empty mapping":
> Unfortunately, the internal structure of `sops` (`TreeBranch`) cannot distinguish "no document" and "empty mapping".

Exactly. The need to represent "no document" seems to be more important than "empty mapping" - see https://github.com/mozilla/sops/issues/695#issuecomment-728263108. I haven't seen (or at least can't remember :) ) any request yet for being able to recover `{}`.

> I think we need to extend `TreeBranch` to have a flag that indicates "no document".

I don't think it can be solved that simply. While this works when cycling between YAML and the internal data structures, it does not help at all when a YAML file is encrypted. Adjusting the metadata won't help since that will not work for multi-document YAML files (metadata is only there once per YAML file).